ygot/struct_validation_map.go

```diff
@@ -319,3 +321,189 @@ func MergeJSON(a, b map[string]interface{}) (map[string]interface{}, error) {
+//
+// TODO(robjs): Fix the unimplemented test cases where two structs of
+// the same type have slices or maps that are already populated.
+// See https://github.com/openconfig/ygot/issues/74.
+func MergeStructs(a, b ValidatedGoStruct) (ValidatedGoStruct, error) {
+	if reflect.TypeOf(a) != reflect.TypeOf(b) {
+		return nil, fmt.Errorf("cannot merge structs that are not of matching types, %T != %T", a, b)
+	}
+
+	n := reflect.New(reflect.TypeOf(a).Elem())
+	if err := copyStruct(n.Elem(), reflect.ValueOf(a).Elem()); err != nil {
+		return nil, fmt.Errorf("error merging a to new struct: %v", err)
+	}
+
+	if err := copyStruct(n.Elem(), reflect.ValueOf(b).Elem()); err != nil {
+		return nil, fmt.Errorf("error merging b to new struct: %v", err)
+	}
+
+	return n.Interface().(ValidatedGoStruct), nil
+
+}
+
+// copyStruct copies the fields of srcVal into the dstVal struct in-place.
+func copyStruct(dstVal, srcVal reflect.Value) error {
+	if srcVal.Type() != dstVal.Type() {
+		return fmt.Errorf("cannot copy %s to %s", srcVal.Type().Name(), dstVal.Type().Name())
+	}
+
+	if !util.IsValueStruct(dstVal) || !util.IsValueStruct(srcVal) {
+		return fmt.Errorf("cannot handle non-struct types, src: %v, dst: %v", srcVal.Type().Kind(), dstVal.Type().Kind())
+	}
+
+	for i := 0; i < srcVal.NumField(); i++ {
+		srcField := srcVal.Field(i)
+		dstField := dstVal.Field(i)
+
+		switch srcField.Kind() {
+		case reflect.Ptr:
+			if err := copyPtrField(dstField, srcField); err != nil {
+				return err
+			}
+		case reflect.Interface:
+			if err := copyInterfaceField(dstField, srcField); err != nil {
+				return err
+			}
+		case reflect.Map:
+			if err := copyMapField(dstField, srcField); err != nil {
+				return err
+			}
+		case reflect.Slice:
+			if err := copySliceField(dstField, srcField); err != nil {
+				return err
+			}
+		default:
+			dstField.Set(srcField)
+		}
+	}
+	return nil
+}
+
+// copyPtrField copies srcField to dstField. srcField and dstField must be
+// reflect.Value structs which represent pointers.
+func copyPtrField(dstField, srcField reflect.Value) error {
+	if srcField.IsNil() || !srcField.IsValid() {
+		return nil
+	}
+
+	switch srcField.Elem().Kind() {
+	case reflect.Struct:
+		var d reflect.Value
+		d = reflect.New(srcField.Type().Elem())
+		if err := copyStruct(d.Elem(), srcField.Elem()); err != nil {
+			return err
+		}
+		dstField.Set(d)
+	default:
+		p := reflect.New(srcField.Type().Elem())
+		p.Elem().Set(srcField.Elem())
+		dstField.Set(p)
+	}
+
+	return nil
+}
+
+// copyInterfaceField copies srcField into dstField. Both srcField and dstField
+// are reflect.Value structs which contain an interface value.
+func copyInterfaceField(dstField, srcField reflect.Value) error {
+	if srcField.IsNil() || !srcField.IsValid() {
+		return nil
+	}
+
+	if srcField.Elem().Kind() != reflect.Ptr {
+		return fmt.Errorf("invalid interface type received: %T", srcField.Interface())
+	}
+
+	switch srcField.Elem().Elem().Kind() {
```

Comment:
Rather than using srcField.Elem().Elem() throughout, suggest the following:
srcField = srcField.Elem().Elem() (after checks above)
then 
if !util.IsValueStruct(srcField) {
 return error
} 
// default.

Reply:
Done.